src/compress.rs

```diff
@@ -0,0 +1,128 @@
+//! Abstract over different compressors that could be used.
+//!
+//! For the time being, we support lz4 and zstd, but since we want a given
+//! shard file to internally specify which compressor was used, we do not
+//! plan to support extensible compressor support. This could be changed in the
+//! future through the use of a compile-time plugin registry system like the
+//! inventory crate.
+
+use std::io::{BufReader, Read, Write};
+
+use anyhow::{bail, Result};
+use zstd::zstd_safe;
+
+/// The compressors supported by shardio.
+#[derive(Clone, Copy)]
+pub enum Compressor {
+    /// lz4 compression at level 2
+    /// this appears to be a good general trade-off of speed and compression ratio.
+    /// note see these benchmarks for useful numers:
+    /// http://quixdb.github.io/squash-benchmark/#ratio-vs-compression
+    /// important note: lz4f (not lz4) is the relevant mode in those charts.
+    Lz4,
+    /// zstd compression using --fast=1
+    /// Roughly the same speed as lz4 level 2 but possibly 40% smaller for certain
+    /// workloads.
+    Zstd,
+}
```

Comment:
It would be easy enough to make the compression level configurable in here.  You don't need to encode it into the file, because the decompressor doesn't care what compression level you use.
```suggestion
#[derive(Clone, Copy)]
pub enum Compressor {
    /// lz4 compression
    /// this appears to be a good general trade-off of speed and compression ratio.
    /// note see these benchmarks for useful numers:
    /// http://quixdb.github.io/squash-benchmark/#ratio-vs-compression
    /// important note: lz4f (not lz4) is the relevant mode in those charts.
    Lz4(u32),
    /// zstd compression
    /// Fast 1 (level -1) is roughly the same speed as lz4 level 2 but possibly 40% smaller for certain
    /// workloads.
    Zstd(i32),
}
```

Providing an appropriate `default` would probably be a good idea though.

Reply:
Yeah, I like the idea of setting the level in this enum and having helper functions for default levels such as `fast, max` etc. But this could be implemented later